File: idea-plugin/src/main/java/com/zzf/rikki/session/SessionProcessor.java
package com.zzf.rikki.session;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.zzf.rikki.core.tool.PendingChangesManager;
import com.zzf.rikki.idea.agent.compat.AgentEventSink;
import com.zzf.rikki.idea.agent.compat.LlmChatRequest;
import com.zzf.rikki.idea.agent.compat.LlmStreamListener;
import com.zzf.rikki.idea.agent.compat.LlmStreamResult;
import com.zzf.rikki.idea.agent.compat.ModelCapabilities;
import com.zzf.rikki.idea.agent.compat.PendingCommandRecord;
import com.zzf.rikki.idea.agent.compat.PendingCommandRegistration;
import com.zzf.rikki.idea.agent.compat.RuntimeEvent;
import com.zzf.rikki.idea.agent.compat.ToolCallInfo;
import com.zzf.rikki.idea.agent.compat.ToolExecutionResult;
import com.zzf.rikki.idea.agent.tools.LiteBashTool;
import com.zzf.rikki.llm.LLMService;
import com.zzf.rikki.runtime.port.PendingApprovalPort;
import com.zzf.rikki.runtime.port.RuntimeRequest;
import com.zzf.rikki.runtime.port.ToolExecutorPort;
import com.zzf.rikki.session.model.MessageV2;
import com.zzf.rikki.session.model.PromptPart;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class SessionProcessor {
    private static final int MAX_TOOL_OUTPUT = 8000;

    public static final class ProcessorResult {
        public final String answer;
        public final String messageId;
        public final boolean continueLoop;

        public ProcessorResult(String answer, String messageId, boolean continueLoop) {
            this.answer = answer;
            this.messageId = messageId;
            this.continueLoop = continueLoop;
        }
    }

    private final ObjectMapper mapper;
    private final SessionService sessionService;
    private final SessionStatus sessionStatus;
    private final PendingApprovalPort pendingApprovalService;
    private final LLMService llmService;
    private final ToolExecutorPort toolExecutor;

    public SessionProcessor(
            ObjectMapper mapper,
            SessionService sessionService,
            SessionStatus sessionStatus,
            PendingApprovalPort pendingApprovalService,
            LLMService llmService,
            ToolExecutorPort toolExecutor
    ) {
        this.mapper = mapper;
        this.sessionService = sessionService;
        this.sessionStatus = sessionStatus;
        this.pendingApprovalService = pendingApprovalService;
        this.llmService = llmService;
        this.toolExecutor = toolExecutor;
    }

    public ProcessorResult process(
            RuntimeRequest request,
            SessionInfo session,
            MessageV2.WithParts assistantMessage,
            List<MessageV2.WithParts> promptMessages,
            String systemPrompt,
            ModelCapabilities capabilities,
            List<Map<String, Object>> toolDefinitions,
            AgentEventSink sink
    ) {
        MessageV2.StepStartPart stepStart = new MessageV2.StepStartPart();
        stepStart.id = nextId("part");
        stepStart.sessionID = session.id;
        stepStart.messageID = assistantMessage.info.id;
        sessionService.updatePart(stepStart);

        final MessageV2.TextPart[] currentText = {null};
        final MessageV2.ReasoningPart[] currentReasoning = {null};

        LlmChatRequest llmRequest = new LlmChatRequest(
                assistantMessage.info.id,
                castMaps(sessionService.toLlmMessages(promptMessages, systemPrompt, capabilities.getSystemRole())),
                capabilities,
                castMaps(toolDefinitions)
        );
        LlmStreamResult llmResult = llmService.streamChat(llmRequest, new LlmStreamListener() {
            @Override
            public void onMessageDelta(String messageId, String delta) {
                if (delta == null || delta.isEmpty()) {
                    return;
                }
                if (currentText[0] == null) {
                    MessageV2.TextPart part = new MessageV2.TextPart();
                    part.id = nextId("part");
                    part.sessionID = session.id;
                    part.messageID = messageId;
                    part.time.start = System.currentTimeMillis();
                    currentText[0] = part;
                    sessionService.updatePart(part);
                }
                currentText[0].delta = delta;
                currentText[0].text += delta;
                currentText[0].time.end = System.currentTimeMillis();
                sessionService.updatePart(currentText[0]);
                sink.emit(new RuntimeEvent.MessageDelta(messageId, delta));
            }

            @Override
            public void onThoughtDelta(String messageId, String delta) {
                if (delta == null || delta.isEmpty()) {
                    return;
                }
                if (currentReasoning[0] == null) {
                    MessageV2.ReasoningPart part = new MessageV2.ReasoningPart();
                    part.id = nextId("part");
                    part.sessionID = session.id;
                    part.messageID = messageId;
                    part.time.start = System.currentTimeMillis();
                    currentReasoning[0] = part;
                    sessionService.updatePart(part);
                }
                currentReasoning[0].delta = delta;
                currentReasoning[0].text += delta;
                currentReasoning[0].time.end = System.currentTimeMillis();
                sessionService.updatePart(currentReasoning[0]);
                sink.emit(new RuntimeEvent.ThoughtDelta(messageId, delta));
            }

            @Override
            public void onThoughtEnd(String messageId) {
                if (currentReasoning[0] != null) {
                    currentReasoning[0].time.end = System.currentTimeMillis();
                    sessionService.updatePart(currentReasoning[0]);
                }
                sink.emit(new RuntimeEvent.ThoughtEnd(messageId));
            }
        });

        if (llmResult.getReasoningContent() != null && !llmResult.getReasoningContent().isBlank() && currentReasoning[0] == null) {
            MessageV2.ReasoningPart part = new MessageV2.ReasoningPart();
            part.id = nextId("part");
            part.sessionID = session.id;
            part.messageID = assistantMessage.info.id;
            part.text = llmResult.getReasoningContent();
            part.delta = llmResult.getReasoningContent();
            part.time.start = System.currentTimeMillis();
            part.time.end = part.time.start;
            sessionService.updatePart(part);
            sink.emit(new RuntimeEvent.ThoughtDelta(assistantMessage.info.id, llmResult.getReasoningContent()));
            sink.emit(new RuntimeEvent.ThoughtEnd(assistantMessage.info.id));
        }

        boolean hasToolCalls = llmResult.getToolCalls() != null && !llmResult.getToolCalls().isEmpty();
        String finishReason = hasToolCalls ? "tool-calls" : "stop";
        MessageV2.StepFinishPart stepFinish = new MessageV2.StepFinishPart();
        stepFinish.id = nextId("part");
        stepFinish.sessionID = session.id;
        stepFinish.messageID = assistantMessage.info.id;
        stepFinish.reason = finishReason;
        stepFinish.tokens = assistantMessage.info.tokens;
        stepFinish.cost = assistantMessage.info.cost;
        sessionService.updatePart(stepFinish);

        if (llmResult.getToolCalls() != null) {
            for (ToolCallInfo toolCall : llmResult.getToolCalls()) {
                Map<String, Object> argsMap = toArgsMap(toolCall.getArgs());
                MessageV2.ToolPart toolPart = new MessageV2.ToolPart();
                toolPart.id = nextId("part");
                toolPart.sessionID = session.id;
                toolPart.messageID = assistantMessage.info.id;
                toolPart.callID = toolCall.getId();
                toolPart.tool = toolCall.getName();
                toolPart.args.putAll(argsMap);
                toolPart.state.status = toolExecutor.isHighRisk(toolCall.getName(), toolCall.getArgs()) ? "pending" : "running";
                toolPart.state.input.putAll(argsMap);
                toolPart.state.title = toolCall.getName();
                toolPart.state.time.start = System.currentTimeMillis();
                sessionService.updatePart(toolPart);
                if (toolExecutor.isHighRisk(toolCall.getName(), toolCall.getArgs())) {
                    handlePendingToolCall(request, session, assistantMessage, toolPart, toolCall, argsMap, sink);
                } else {
                    sink.emit(new RuntimeEvent.ToolCall(
                            toolPart.id,
                            toolPart.tool,
                            toolPart.callID,
                            assistantMessage.info.id,
                            "running",
                            toolPart.tool,
                            argsMap,
                            null
                    ));
                    ToolExecutionResult result = toolExecutor.execute(
                            toolCall.getName(),
                            toolCall.getArgs(),
                            request.getWorkspaceRoot(),
                            session.id,
                            toolCall.getId(),
                            assistantMessage.info.id
                    );
                    applyToolResult(session, toolPart, result, sink);
                }
            }
        }

        if (!hasToolCalls && llmResult.getText() != null && llmResult.getText().startsWith("Error:")) {
            assistantMessage.info.error = new MessageV2.ErrorInfo();
            assistantMessage.info.error.message = llmResult.getText();
            assistantMessage.info.error.type = "llm_error";
            assistantMessage.info.finishReason = "error";
        } else {
            assistantMessage.info.finishReason = finishReason;
        }
        assistantMessage.info.finish = Boolean.TRUE;
        assistantMessage.info.time.end = System.currentTimeMillis();
        sessionService.updateMessage(assistantMessage);

        return new ProcessorResult(llmResult.getText(), assistantMessage.info.id, hasToolCalls);
    }

    private void handlePendingToolCall(
            RuntimeRequest request,
            SessionInfo session,
            MessageV2.WithParts assistantMessage,
            MessageV2.ToolPart toolPart,
            ToolCallInfo toolCall,
            Map<String, Object> argsMap,
            AgentEventSink sink
    ) {
        PendingCommandRecord pendingRecord = buildPendingCommandRecord(toolCall, argsMap, request.getWorkspaceRoot(), session.id, assistantMessage.info.id);
        PendingCommandRegistration registration = pendingApprovalService.registerPendingCommand(pendingRecord);
        Map<String, Object> meta = pendingMeta(registration.getPendingCommand());
        sink.emit(new RuntimeEvent.ToolCall(
                toolPart.id,
                toolPart.tool,
                toolPart.callID,
                assistantMessage.info.id,
                registration.getAwaitApproval() ? "pending" : "running",
                toolPart.tool,
                argsMap,
                meta
        ));
        toolPart.state.metadata.putAll(meta);
        sessionService.updatePart(toolPart);
        ToolExecutionResult result;
        if (registration.getAwaitApproval()) {
            sessionStatus.set(session.id, new SessionStatus.Info("waiting_approval", null, "Awaiting your approval...", null));
            sink.emit(new RuntimeEvent.StatusChanged("waiting_approval", "Awaiting your approval..."));
            if (registration.getPendingCommand().getStrictApproval()) {
                sink.emit(new RuntimeEvent.ToolPendingApproval(toolPart.id, toolPart.callID, registration.getPendingCommand().getCommand(), toolPart.tool));
            }
            result = pendingApprovalService.awaitDecision(registration.getPendingCommand().getId());
            sessionStatus.set(session.id, new SessionStatus.Info("busy", null, "Agent is thinking...", null));
            sink.emit(new RuntimeEvent.StatusChanged("busy", "Agent is thinking..."));
        } else {
            result = registration.getImmediateResult();
            if (result == null) {
                result = new ToolExecutionResult("error", "", "Auto-approved command result missing.", Map.of(), null, null, null, null, false);
            }
        }
        applyToolResult(session, toolPart, result, sink);
    }

    private void applyToolResult(SessionInfo session, MessageV2.ToolPart toolPart, ToolExecutionResult result, AgentEventSink sink) {
        Long startedAt = toolPart.state.time.start;
        LinkedHashMap<String, Object> meta = new LinkedHashMap<>();
        meta.putAll(result.getMeta());
        List<PromptPart> emittedParts = materializeEmittedParts(toolPart.sessionID, toolPart.messageID, meta.get("emitted_parts"));
        PendingChangesManager.PendingChange pendingChange = result.getPendingChange();
        if (pendingChange != null) {
            meta.put("pending_change", pendingChange);
        }
        PendingCommandRecord pendingCommand = result.getPendingCommand();
        if (pendingCommand != null) {
            meta.put("pending_command", pendingCommand.toMetaMap());
        }
        MessageV2.ToolState nextState = new MessageV2.ToolState();
        nextState.status = result.getStatus() == null || result.getStatus().isBlank() ? "completed" : result.getStatus();
        nextState.input.putAll(toolPart.args);
        nextState.output = result.getOutput() == null ? "" : truncate(result.getOutput(), MAX_TOOL_OUTPUT);
        nextState.title = toolPart.tool;
        nextState.error = result.getError();
        nextState.metadata.putAll(meta);
        nextState.time.start = startedAt;
        nextState.time.end = System.currentTimeMillis();
        toolPart.state = nextState;
        sessionService.updatePart(toolPart);
        for (PromptPart part : emittedParts) {
            sessionService.updatePart(part);
        }
        sink.emit(new RuntimeEvent.ToolResult(
                toolPart.id,
                toolPart.tool,
                toolPart.callID,
                toolPart.messageID,
                toolPart.state.status,
                toolPart.tool,
                toolPart.state.output,
                toolPart.state.error,
                meta.isEmpty() ? null : meta
        ));
        if (result.getTodoJson() != null) {
            sink.emit(new RuntimeEvent.TodoUpdated(result.getTodoJson(), session.id));
        }
    }

    private List<PromptPart> materializeEmittedParts(String sessionId, String messageId, Object rawParts) {
        List<PromptPart> parts = new ArrayList<>();
        if (!(rawParts instanceof List<?> rawList)) {
            return parts;
        }
        for (Object candidate : rawList) {
            try {
                JsonNode node = mapper.valueToTree(candidate);
                PromptPart part = sessionService.deserializePart(sessionId, messageId, node);
                if (part != null) {
                    if (part.sessionID == null || part.sessionID.isBlank()) {
                        part.sessionID = sessionId;
                    }
                    if (part.messageID == null || part.messageID.isBlank()) {
                        part.messageID = messageId;
                    }
                    if (part.id == null || part.id.isBlank()) {
                        part.id = nextId("part");
                    }
                    parts.add(part);
                }
            } catch (Exception ignored) {
            }
        }
        return parts;
    }

    private PendingCommandRecord buildPendingCommandRecord(
            ToolCallInfo toolCall,
            Map<String, Object> argsMap,
            String workspaceRoot,
            String sessionId,
            String messageId
    ) {
        String shell = toolCall.getArgs().path("shell").asText("auto");
        if (shell.isBlank()) {
            shell = "auto";
        }
        long timeoutMs = toolCall.getArgs().path("timeout").asLong(com.zzf.rikki.idea.agent.compat.CommandRunner.DEFAULT_TIMEOUT_MS);
        if (timeoutMs <= 0L) {
            timeoutMs = com.zzf.rikki.idea.agent.compat.CommandRunner.DEFAULT_TIMEOUT_MS;
        }
        String workdir = toolCall.getArgs().path("workdir").asText(workspaceRoot);
        if (workdir.isBlank()) {
            workdir = workspaceRoot;
        }
        String command = "bash".equals(toolCall.getName())
                ? toolCall.getArgs().path("command").asText("(unknown)")
                : "Delete: " + toolCall.getArgs().path("filePath").asText("(unknown)");
        boolean strictApproval = "delete_file".equals(toolCall.getName()) || LiteBashTool.Companion.isStrictApprovalCommand(command);
        String riskCategory = strictApproval ? "destructive" : "restricted";
        List<String> reasons = new ArrayList<>(LiteBashTool.Companion.detectRiskReasons(command));
        if (reasons.isEmpty() && "delete_file".equals(toolCall.getName())) {
            reasons.add("file deletion requires approval");
        }
        String description = argsMap.get("description") instanceof String value ? value : toolCall.getName();
        return new PendingCommandRecord(
                nextId("pending"),
                command,
                description,
                workdir,
                workspaceRoot,
                sessionId,
                timeoutMs,
                toolCall.getName(),
                toolCall.getId(),
                messageId,
                "high",
                riskCategory,
                LiteBashTool.Companion.commandFamily(command),
                strictApproval,
                reasons,
                shell,
                () -> toolExecutor.execute(
                        toolCall.getName(),
                        toolCall.getArgs(),
                        workspaceRoot,
                        sessionId,
                        toolCall.getId(),
                        messageId
                )
        );
    }

    private Map<String, Object> pendingMeta(PendingCommandRecord pendingCommand) {
        List<String> approvalOptions = pendingCommand.getStrictApproval()
                ? List.of("manual")
                : List.of("manual", "whitelist", "always_allow_non_destructive");
        return mapOf(
                "pending_command", pendingCommand.toMetaMap(),
                "approval_options", approvalOptions,
                "requires_explicit_user_consent", Boolean.TRUE,
                "risk_level", pendingCommand.getRiskLevel(),
                "risk_category", pendingCommand.getRiskCategory(),
                "risk_reasons", pendingCommand.getReasons()
        );
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> toArgsMap(JsonNode args) {
        try {
            return mapper.convertValue(args, LinkedHashMap.class);
        } catch (Exception ignored) {
            return Map.of();
        }
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    private static List<Map<String, Object>> castMaps(List<? extends Map<String, ?>> raw) {
        return (List) raw;
    }

    private static LinkedHashMap<String, Object> mapOf(Object... pairs) {
        LinkedHashMap<String, Object> map = new LinkedHashMap<>();
        for (int i = 0; i < pairs.length; i += 2) {
            map.put((String) pairs[i], pairs[i + 1]);
        }
        return map;
    }

    private static String truncate(String text, int max) {
        if (text == null || text.length() <= max) {
            return text == null ? "" : text;
        }
        return text.substring(0, max);
    }

    private static String nextId(String prefix) {
        return prefix + "-" + UUID.randomUUID();
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/compat/SessionProcessor.kt
package com.zzf.rikki.idea.agent.compat

import com.fasterxml.jackson.databind.ObjectMapper
import com.zzf.rikki.idea.agent.tools.LiteBashTool
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

private const val MAX_TOOL_OUTPUT = 8_000

data class ProcessorResult(
    val answer: String,
    val messageId: String,
    val continueLoop: Boolean
)

class SessionProcessor(
    private val mapper: ObjectMapper,
    private val sessionService: SessionService,
    private val sessionStatus: SessionStatus,
    private val pendingApprovalService: InMemoryPendingApprovalService,
    private val llmStreamClient: LlmStreamClient,
    private val toolExecutor: ToolExecutor
) {
    suspend fun process(
        request: ChatRuntimeRequest,
        session: SessionInfo,
        assistantMessage: MessageWithParts,
        systemPrompt: String,
        capabilities: ModelCapabilities,
        toolDefinitions: List<Map<String, Any>>,
        sink: AgentEventSink
    ): ProcessorResult {
        val stepStart = StepStartPart(
            id = nextRuntimeId("part"),
            sessionID = session.id,
            messageID = assistantMessage.info.id
        )
        sessionService.updatePart(stepStart)

        var currentText: TextPart? = null
        var currentReasoning: ReasoningPart? = null
        val llmResult = llmStreamClient.streamChat(
            request = LlmChatRequest(
                messageId = assistantMessage.info.id,
                messages = sessionService.toLlmMessages(session.id, systemPrompt, capabilities.systemRole),
                capabilities = capabilities,
                toolDefinitions = toolDefinitions
            ),
            listener = object : LlmStreamListener {
                override fun onMessageDelta(messageId: String, delta: String) {
                    if (delta.isEmpty()) {
                        return
                    }
                    currentText = currentText ?: TextPart(
                        id = nextRuntimeId("part"),
                        sessionID = session.id,
                        messageID = messageId,
                        time = PartTime(start = System.currentTimeMillis())
                    ).also {
                        sessionService.updatePart(it)
                    }
                    currentText?.let { part ->
                        part.delta = delta
                        part.text += delta
                        part.time?.end = System.currentTimeMillis()
                        sessionService.updatePart(part)
                    }
                    sink.emit(RuntimeEvent.MessageDelta(messageId, delta))
                }

                override fun onThoughtDelta(messageId: String, delta: String) {
                    if (delta.isEmpty()) {
                        return
                    }
                    currentReasoning = currentReasoning ?: ReasoningPart(
                        id = nextRuntimeId("part"),
                        sessionID = session.id,
                        messageID = messageId,
                        time = PartTime(start = System.currentTimeMillis())
                    ).also {
                        sessionService.updatePart(it)
                    }
                    currentReasoning?.let { part ->
                        part.delta = delta
                        part.text += delta
                        part.time?.end = System.currentTimeMillis()
                        sessionService.updatePart(part)
                    }
                    sink.emit(RuntimeEvent.ThoughtDelta(messageId, delta))
                }

                override fun onThoughtEnd(messageId: String) {
                    currentReasoning?.let {
                        it.time?.end = System.currentTimeMillis()
                        sessionService.updatePart(it)
                    }
                    sink.emit(RuntimeEvent.ThoughtEnd(messageId))
                }
            }
        )

        if (llmResult.reasoningContent.isNotBlank() && currentReasoning == null) {
            val reasoningPart = ReasoningPart(
                id = nextRuntimeId("part"),
                sessionID = session.id,
                messageID = assistantMessage.info.id,
                text = llmResult.reasoningContent,
                delta = llmResult.reasoningContent,
                time = PartTime(start = System.currentTimeMillis(), end = System.currentTimeMillis())
            )
            sessionService.updatePart(reasoningPart)
            sink.emit(RuntimeEvent.ThoughtDelta(assistantMessage.info.id, llmResult.reasoningContent))
            sink.emit(RuntimeEvent.ThoughtEnd(assistantMessage.info.id))
        }

        val finishReason = if (llmResult.toolCalls.isEmpty()) "stop" else "tool-calls"
        val stepFinish = StepFinishPart(
            id = nextRuntimeId("part"),
            sessionID = session.id,
            messageID = assistantMessage.info.id,
            reason = finishReason,
            tokens = assistantMessage.info.tokens,
            cost = assistantMessage.info.cost
        )
        sessionService.updatePart(stepFinish)

        for (toolCall in llmResult.toolCalls) {
            currentCoroutineContext().ensureActive()
            val argsMap = LinkedHashMap(toArgsMap(toolCall.args))
            val toolPart = ToolPart(
                id = nextRuntimeId("part"),
                sessionID = session.id,
                messageID = assistantMessage.info.id,
                callID = toolCall.id,
                tool = toolCall.name,
                args = argsMap,
                state = ToolState(
                    status = if (toolExecutor.isHighRisk(toolCall.name, toolCall.args)) "pending" else "running",
                    input = LinkedHashMap(argsMap),
                    title = toolCall.name,
                    time = ToolStateTimeInfo(start = System.currentTimeMillis())
                )
            )
            sessionService.updatePart(toolPart)
            if (toolExecutor.isHighRisk(toolCall.name, toolCall.args)) {
                handlePendingToolCall(request, session, assistantMessage, toolPart, toolCall, argsMap, sink)
            } else {
                sink.emit(
                    RuntimeEvent.ToolCall(
                        partId = toolPart.id,
                        tool = toolPart.tool,
                        callId = toolPart.callID,
                        messageId = assistantMessage.info.id,
                        state = "running",
                        title = toolPart.tool,
                        args = argsMap
                    )
                )
                val result = toolExecutor.execute(
                    name = toolCall.name,
                    args = toolCall.args,
                    workspaceRoot = request.workspaceRoot,
                    sessionId = session.id,
                    callId = toolCall.id,
                    messageId = assistantMessage.info.id
                )
                applyToolResult(session, toolPart, result, sink)
            }
        }

        if (llmResult.text.startsWith("Error:") && llmResult.toolCalls.isEmpty()) {
            assistantMessage.info.error = ErrorInfo(message = llmResult.text, type = "llm_error")
            assistantMessage.info.finishReason = "error"
        } else {
            assistantMessage.info.finishReason = finishReason
        }
        assistantMessage.info.finish = true
        assistantMessage.info.time.end = System.currentTimeMillis()
        sessionService.updateMessage(assistantMessage)

        return ProcessorResult(
            answer = llmResult.text,
            messageId = assistantMessage.info.id,
            continueLoop = llmResult.toolCalls.isNotEmpty()
        )
    }

    private suspend fun handlePendingToolCall(
        request: ChatRuntimeRequest,
        session: SessionInfo,
        assistantMessage: MessageWithParts,
        toolPart: ToolPart,
        toolCall: ToolCallInfo,
        argsMap: Map<String, Any?>,
        sink: AgentEventSink
    ) {
        val pendingRecord = buildPendingCommandRecord(
            toolCall = toolCall,
            argsMap = argsMap,
            workspaceRoot = request.workspaceRoot,
            sessionId = session.id,
            messageId = assistantMessage.info.id
        )
        val registration = pendingApprovalService.registerPendingCommand(pendingRecord)
        val meta = pendingMeta(registration.pendingCommand)
        sink.emit(
            RuntimeEvent.ToolCall(
                partId = toolPart.id,
                tool = toolPart.tool,
                callId = toolPart.callID,
                messageId = assistantMessage.info.id,
                state = if (registration.awaitApproval) "pending" else "running",
                title = toolPart.tool,
                args = argsMap,
                meta = meta
            )
        )
        toolPart.state.metadata.putAll(meta)
        sessionService.updatePart(toolPart)
        if (registration.awaitApproval) {
            sessionStatus.set(session.id, SessionStatusInfo(type = "waiting_approval", message = "Awaiting your approval..."))
            sink.emit(RuntimeEvent.StatusChanged("waiting_approval", "Awaiting your approval..."))
            if (registration.pendingCommand.strictApproval) {
                sink.emit(
                    RuntimeEvent.ToolPendingApproval(
                        partId = toolPart.id,
                        callId = toolPart.callID,
                        command = registration.pendingCommand.command,
                        tool = toolPart.tool
                    )
                )
            }
            val result = pendingApprovalService.awaitDecision(registration.pendingCommand.id)
            sessionStatus.set(session.id, SessionStatusInfo(type = "busy", message = "Agent is thinking..."))
            sink.emit(RuntimeEvent.StatusChanged("busy", "Agent is thinking..."))
            applyToolResult(session, toolPart, result, sink)
        } else {
            val result = registration.immediateResult ?: ToolExecutionResult(
                status = "error",
                output = "",
                error = "Auto-approved command result missing."
            )
            applyToolResult(session, toolPart, result, sink)
        }
    }

    private fun applyToolResult(
        session: SessionInfo,
        toolPart: ToolPart,
        result: ToolExecutionResult,
        sink: AgentEventSink
    ) {
        val startedAt = toolPart.state.time.start
        val meta = LinkedHashMap<String, Any?>()
        meta.putAll(result.meta)
        result.pendingChange?.let { meta["pending_change"] = it }
        result.pendingCommand?.let { meta["pending_command"] = it.toMetaMap() }
        toolPart.state = ToolState(
            status = if (result.status.isBlank()) "completed" else result.status,
            input = LinkedHashMap(toolPart.args),
            output = result.output.take(MAX_TOOL_OUTPUT),
            title = toolPart.tool,
            error = result.error,
            metadata = meta,
            time = ToolStateTimeInfo(start = startedAt, end = System.currentTimeMillis())
        )
        sessionService.updatePart(toolPart)
        sink.emit(
            RuntimeEvent.ToolResult(
                partId = toolPart.id,
                tool = toolPart.tool,
                callId = toolPart.callID,
                messageId = toolPart.messageID,
                state = toolPart.state.status,
                title = toolPart.tool,
                output = toolPart.state.output,
                error = toolPart.state.error,
                meta = if (meta.isEmpty()) null else meta
            )
        )
        if (result.todoJson != null) {
            sink.emit(RuntimeEvent.TodoUpdated(result.todoJson, session.id))
        }
    }

    private fun toArgsMap(args: com.fasterxml.jackson.databind.JsonNode): Map<String, Any?> = try {
        @Suppress("UNCHECKED_CAST")
        mapper.convertValue(args, Map::class.java) as Map<String, Any?>
    } catch (_: Exception) {
        emptyMap()
    }

    private fun pendingMeta(pendingCommand: PendingCommandRecord): Map<String, Any?> = linkedMapOf(
        "pending_command" to pendingCommand.toMetaMap(),
        "approval_options" to if (pendingCommand.strictApproval) {
            listOf(PendingApprovalService.DECISION_MANUAL)
        } else {
            listOf(
                PendingApprovalService.DECISION_MANUAL,
                PendingApprovalService.DECISION_WHITELIST,
                PendingApprovalService.DECISION_ALWAYS_ALLOW_NON_DESTRUCTIVE
            )
        },
        "requires_explicit_user_consent" to true,
        "risk_level" to pendingCommand.riskLevel,
        "risk_category" to pendingCommand.riskCategory,
        "risk_reasons" to pendingCommand.reasons
    )

    private fun buildPendingCommandRecord(
        toolCall: ToolCallInfo,
        argsMap: Map<String, Any?>,
        workspaceRoot: String,
        sessionId: String,
        messageId: String
    ): PendingCommandRecord {
        val shell = toolCall.args.path("shell").asText("auto").ifBlank { "auto" }
        val timeoutMs = toolCall.args.path("timeout").asLong(CommandRunner.DEFAULT_TIMEOUT_MS).let {
            if (it <= 0L) CommandRunner.DEFAULT_TIMEOUT_MS else it
        }
        val workdir = toolCall.args.path("workdir").asText(workspaceRoot).ifBlank { workspaceRoot }
        val command = if (toolCall.name == "bash") {
            toolCall.args.path("command").asText("(unknown)")
        } else {
            "Delete: ${toolCall.args.path("filePath").asText("(unknown)")}"
        }
        val strictApproval = toolCall.name == "delete_file" || LiteBashTool.isStrictApprovalCommand(command)
        val riskCategory = if (strictApproval) "destructive" else "restricted"
        val reasons = LiteBashTool.detectRiskReasons(command).ifEmpty {
            if (toolCall.name == "delete_file") listOf("file deletion requires approval") else emptyList()
        }
        return PendingCommandRecord(
            id = nextRuntimeId("pending"),
            command = command,
            description = argsMap["description"] as? String ?: toolCall.name,
            workdir = workdir,
            workspaceRoot = workspaceRoot,
            sessionId = sessionId,
            timeoutMs = timeoutMs,
            tool = toolCall.name,
            callId = toolCall.id,
            messageId = messageId,
            riskLevel = "high",
            riskCategory = riskCategory,
            commandFamily = LiteBashTool.commandFamily(command),
            strictApproval = strictApproval,
            reasons = reasons,
            shell = shell,
            executor = {
                toolExecutor.execute(
                    name = toolCall.name,
                    args = toolCall.args,
                    workspaceRoot = workspaceRoot,
                    sessionId = sessionId,
                    callId = toolCall.id,
                    messageId = messageId
                )
            }
        )
    }
}
